misc-scripts/xref_mapping/XrefParser/MIMParser.pm

```diff
@@ -21,162 +21,245 @@ package XrefParser::MIMParser;
 
 use strict;
 use warnings;
+
 use Carp;
-use POSIX qw(strftime);
 use File::Basename;
+use Readonly;
+use POSIX qw(strftime);
+
+use parent qw( XrefParser::BaseParser );
+
+# This parser will read xrefs from a record file downloaded from the
+# OMIM Web site. They should be assigned to two different xref
+# sources: MIM_GENE and MIM_MORBID. MIM xrefs are linked to EntrezGene
+# entries so the parser does not match them to Ensembl; this will be
+# taken care of when EntrezGene antries are matched.
+#
+# OMIM records are multiline. Each record begins with a specific tag
+# line and consists of a number of fields. Each field starts with its
+# own start-tag line (i.e. the data proper only appears after a
+# newline) and continues until the beginning of either the next field
+# in the same record, the next record, or the end-of-input tag. The
+# overall structure looks as follows:
+#
+#   *RECORD*
+#   *FIELD* NO
+#   *FIELD* TI
+#   *FIELD* TX
+#   ...
+#   *RECORD*
+#   *FIELD* NO
+#   *FIELD* TI
+#   ...
+#   *RECORD*
+#   *FIELD* NO
+#   ...
+#   *FIELD* CD
+#   *FIELD* ED
+#   *THEEND*
+#
+# All the data relevant to the parser can be found in the TI field.
+
+
+# FIXME: this belongs in BaseParser
+Readonly my $ERR_SOURCE_ID_NOT_FOUND => -1;
 
-use base qw( XrefParser::BaseParser );
 
 sub run {
 
-  my ($self, $ref_arg) = @_;
-  my $general_source_id    = $ref_arg->{source_id};
-  my $species_id   = $ref_arg->{species_id};
-  my $files        = $ref_arg->{files};
-  my $verbose      = $ref_arg->{verbose};
-  my $dbi          = $ref_arg->{dbi};
-  $dbi = $self->dbi unless defined $dbi;
+  my ( $self, $ref_arg ) = @_;
+  my $general_source_id = $ref_arg->{source_id};
+  my $species_id        = $ref_arg->{species_id};
+  my $files             = $ref_arg->{files};
+  my $verbose           = $ref_arg->{verbose};
+  my $dbi               = $ref_arg->{dbi};
 
-  if((!defined $general_source_id) or (!defined $species_id) or (!defined $files) ){
+  if ( ( !defined $general_source_id ) or
+       ( !defined $species_id ) or
+       ( !defined $files ) )
+  {
     croak "Need to pass source_id, species_id and files as pairs";
```

Comment:
I don't think croak is giving us the errors we desire. For me it was "Error in run() in file", rather than "Error on line n"

Reply:
Hmm, all the four parsers I have been assigned handle this particular bit identically _i.e._ with croak()... Would you have any suggestions regarding how to improve this? By the way, this test never gets triggered if _standaloneJob.pl_ is used, one gets _e.g._

` WORKER_ERROR : ParamError: value for param_required('source') is required and has to be defined`

instead.

Reply:
The ParseSource runnable checks for required parameters, see https://github.com/Ensembl/ensembl-production/blob/release/94/modules/Bio/EnsEMBL/Production/Pipeline/Xrefs/ParseSource.pm#L30
It is probably good to leave the croak for backwards compatibility with the xref_parser.pl script, but also comforting to know the eHive pipeline does a better job at checking for input

Reply:
OK